### PR TITLE
[3.x] Improve performance of editor grid redrawing

### PR DESCRIPTION
**Pull request by RevoluPowered**
This was noticed in the MacOS profiler and will apply to other platforms too:

[optimized before]
Screenshot 2021-08-06 at 23 51 19

[optimized after]
Screenshot 2021-08-06 at 23 52 49

[unoptimized before]
Screenshot 2021-08-06 at 23 04 49

[unoptimized after]
Screenshot 2021-08-06 at 23 08 55

Most of us are deving on debug so will be nice to not have this taking up a lot of CPU.

### CODE REVIEW

**Review by clayjohn — COMMENTED** [review]
Woohoo a free millisecond every frame is always welcome

**Comment by RevoluPowered** [comment]
latest improvement: significantly reduces cpu time spent on this task.

Screenshot 2021-08-07 at 03 48 32

**Comment by Chaosus** [comment]
Can this be ported to master branch?

**Comment by RevoluPowered** [comment]
> Can this be ported to master branch?

Yes, but I need to see if it has any discernible benefit :) one sec I'll check per figures with/without

**Comment by RevoluPowered** [comment]
Screenshot 2021-08-07 at 17 06 25

This is the single process in 4.0 occupying a lot of CPU time so yes it will transfer over PR incoming :D

**Comment by RevoluPowered** [comment]
4.0 PR is made and perf saving is awesome: #51399

File: editor/plugins/spatial_editor_plugin.cpp [hunk 1]
```diff
@@ -5674,6 +5682,32 @@ void SpatialEditor::_init_grid() {
 		grid_mat[c]->set_shader_param("grid_size", grid_fade_size);
 		grid_mat[c]->set_shader_param("orthogonal", orthogonal);
 
+		// cache these so we don't have to re-access memory.
```
Comment (LightningAA) [inline editor/plugins/spatial_editor_plugin.cpp:5685]:
```suggestion
		// Cache these so we don't have to re-access memory.
```
https://docs.godotengine.org/en/stable/community/contributing/code_style_guidelines.html#comment-style-guide

File: editor/plugins/spatial_editor_plugin.cpp [hunk 2]
```diff
@@ -5751,10 +5787,22 @@ void SpatialEditor::_finish_grid() {
 		VisualServer::get_singleton()->free(grid[i]);
 	}
 }
-
+bool setup = false;
+Vector3 cumulativeMovement = Vector3();
```
Comment (LightningAA) [inline editor/plugins/spatial_editor_plugin.cpp:5791]:
```suggestion
Vector3 cumulative_movement = Vector3();
```

**Review by AaronRecord — CHANGES_REQUESTED** [review]

File: editor/plugins/spatial_editor_plugin.cpp (hunk 2 above)
Comment (RevoluPowered) [reply editor/plugins/spatial_editor_plugin.cpp]: Fixed

File: editor/plugins/spatial_editor_plugin.cpp (hunk 1 above)
Comment (RevoluPowered) [reply editor/plugins/spatial_editor_plugin.cpp:5685]: Fixed

**Comment by RevoluPowered** [comment]
TODO: Improve grid forward/backward redrawing

File: editor/plugins/spatial_editor_plugin.cpp (hunk 1 above)
Comment (LightningAA) [reply editor/plugins/spatial_editor_plugin.cpp:5685]: ... doesn't look fixed yet 🙂

**Comment by RevoluPowered** [comment]
4.0 PR has an orthographic fix which should be ported

**Comment by RevoluPowered** [comment]
closing this can just be cherrypicked from master

**Pull request closed without merging**